Mathlib/Computability/MyhillNerode.lean

```diff
@@ -0,0 +1,98 @@
+/-
+Copyright (c) 2024 Google. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Wong
+-/
+import Mathlib.Computability.DFA
+import Mathlib.Data.Set.Finite.Basic
+
+/-!
+# Myhill–Nerode theorem
+
+This file proves the Myhill–Nerode theorem using left quotients.
+
+Given a language `L` and a word `x`, the *left quotient* of `L` by `x` is the set of suffixes `y`
+such that `x ++ y` is in `L`. The *Myhill–Nerode theorem* shows that each left quotient, in fact,
+corresponds to the state of an automaton that matches `L`, and that `L` is regular if and only if
+there are finitely many such states.
+
+## References
+
+* <https://en.wikipedia.org/wiki/Syntactic_monoid#Myhill%E2%80%93Nerode_theorem>
+-/
+
+universe u v
+variable {α : Type u} {σ : Type v} (L : Language α)
+
+namespace Language
+
+/-- The *left quotient* of `x` is the set of suffixes `y` such that `x ++ y` is in `L`. -/
+def leftQuotient (x : List α) : Language α := { y | x ++ y ∈ L }
+
+@[simp]
+theorem leftQuotient_nil : L.leftQuotient [] = L := rfl
+
+theorem leftQuotient_append (x y : List α) :
+    L.leftQuotient (x ++ y) = (L.leftQuotient x).leftQuotient y := by
+  simp [leftQuotient, Language]
+
+@[simp]
+theorem mem_leftQuotient (x y : List α) : y ∈ L.leftQuotient x ↔ x ++ y ∈ L := Iff.rfl
+
+theorem leftQuotient_accepts (M : DFA α σ) (x : List α) :
+    leftQuotient M.accepts x = M.acceptsFrom (M.eval x) := by
+  ext y
+  simp [DFA.mem_accepts, DFA.mem_acceptsFrom, DFA.eval, DFA.evalFrom_of_append]
+
+/-- Like `leftQuotient_accepts`, but as an equality on functions. -/
+theorem leftQuotient_accepts' (M : DFA α σ) : leftQuotient M.accepts = M.acceptsFrom ∘ M.eval :=
```

Comment:
Usually we differentiate between the applied and unapplied versions by suffixing `_apply`, eg
```suggestion
theorem leftQuotient_accepts_apply (M : DFA α σ) (x : List α) :
    leftQuotient M.accepts x = M.acceptsFrom (M.eval x) := by
  ext y
  simp [DFA.mem_accepts, DFA.mem_acceptsFrom, DFA.eval, DFA.evalFrom_of_append]

theorem leftQuotient_accepts (M : DFA α σ) : leftQuotient M.accepts = M.acceptsFrom ∘ M.eval :=
```